Restore highest producer ID during cluster recovery

andrwng opened the pull request:
This PR restores the highest producer ID used by any uploaded segment data in the previous cluster. To accomplish this, there are a few major components:
- At restore time, the highest producer ID is queried across all nodes, loaded from their manifests, sending out an RPC to all nodes to aggregate the highest producer ID of any partition across all shards.
- To make the above step reliable, a sync point is added just after restoring topics from the controller snapshot, using `wait_for_topics()` to wait for partitions to be bootstrapped before proceeding.
- An RPC is then sent to the id_allocator partition to reset the next ID. A new method is used to persist this next ID using the existing STM commands, and the same routing logic is used to direct the requests to the leader.
- Finally, this is plugged into the cluster_recovery_backend before marking the cluster recovery as complete.

Why can't we just upload a single producer ID from the id_allocator when uploading metadata? Segment uploads are not coordinated by any central entity. This means that if there were a central uploader that queried the highest producer ID from the id_allocator, there would be a strong chance that new transactional segments get initialized and uploaded before the next ID is included in uploaded metadata.

This adds a couple fixture tests to validate the end-to-end behavior. Ducktape tests will follow, once the rest of cluster recovery (offsets recovery) is plugged in.

Fixes https://github.com/redpanda-data/redpanda/issues/9555

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng commented:
Pulled out the first couple of commits into its own PR: https://github.com/redpanda-data/redpanda/pull/15056

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/41434#018bee2d-ce7a-4a63-9fc1-444741aeb75f:
```
"rptest.tests.upgrade_test.UpgradeFromPriorFeatureVersionCloudStorageTest.test_rolling_upgrade.cloud_storage_type=CloudStorageType.S3"
```
<!--4920c21f-ea44-4e9c-abe2-dfe44eede894-->

dotnwat commented:
>At restore time, the highest producer ID is queried across all nodes, sending out an RPC to all nodes to aggregate the highest producer ID of any partition across all shards.

are there failure modes here to be concerned with? for example, if this process is idempotent and has a well defined barrier after it completes then presumably any interim failures can just be restarted?

andrwng commented:
> > At restore time, the highest producer ID is queried across all nodes, sending out an RPC to all nodes to aggregate the highest producer ID of any partition across all shards.
> 
> are there failure modes here to be concerned with? for example, if this process is idempotent and has a well defined barrier after it completes then presumably any interim failures can just be restarted?

The restoration is meant to be idempotent, so retrying is typically the way to go. There are timeouts baked in here and there (frankly, should be cleaned up to be more uniform) such that if e.g. it takes Too Long to create some topics, the recovery is marked as failed and someone should probably take a look to see what's up. But the entire cluster recovery can then be retried again.

andrwng commented on src/v/cluster/partition.h:
```diff
@@ -322,7 +322,16 @@ class partition {
     model::producer_id highest_producer_id() {
-        return _rm_stm->highest_producer_id();
+        auto pid = _rm_stm ? _rm_stm->highest_producer_id()
+                           : model::producer_id{};
```
That's right. Can change this to optional (we've historically not been very consistent, but there are some optionals now in the manifest)

dotnwat replied:
yeh no reason to change it just wondering

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -1336,6 +1368,13 @@ void archival_metadata_stm::apply_reset_metadata() {
+void archival_metadata_stm::apply_update_highest_producer_id(
+  model::producer_id pid) {
+    if (_manifest->advance_highest_producer_id(pid)) {
+        vlog(_logger.debug, "Updated highest producer ID to {}", pid());
```
What kind of context are you thinking of? Keep in mind, _logger is a prefix logger so it includes the NTP

dotnwat replied:
>Keep in mind, _logger is a prefix logger

Nevermind!

andrwng commented on src/v/cluster/archival_metadata_stm.cc:
```diff
@@ -897,6 +914,13 @@ ss::future<std::error_code> archival_metadata_stm::do_add_segments(
 
+    if (highest_pid != model::producer_id{}) {
+        iobuf key_buf = serde::to_iobuf(
+          archival_metadata_stm::update_highest_producer_id_cmd::key);
+        iobuf val_buf = serde::to_iobuf(highest_pid());
```
I'm not against it, it's a good practice. But I'm also not sure how this would be extended (and if push comes to shove, at that time, we can always deprecate this and add a new one)

dotnwat replied:
Ill leave it up to you I don't have a strong opinion about it.

andrwng commented on src/v/cluster/partition_leaders_table.cc:
```diff
@@ -240,6 +241,34 @@ ss::future<> partition_leaders_table::update_with_estimates() {
     }
 }
 
+ss::future<partition_leaders_table::leaders_count_t>
+partition_leaders_table::get_leaders_count(
```
Good question. Maybe this is referring to the "estimated" leadership?

mmaslankaprv replied:
This method is not safe to use, with `ss::maybe_yield` in the loop it may lead to a situation in which topic table will change after `maybe_yield` returns leading to a crash.

What is the purpose for this method ? If we need to wait for topic partitions we have an API in `controller_api`

mmaslankaprv replied:
also we are indeed trying to guesstimate leadership when topic is created which may happen before partitions are even present.

andrwng replied:
Note that `topics` is a copy whose lifecycle is tied to this coroutine, which shoudl be safe, though noted it's definitely something to do a double take at and could use a comment...

I'd like to ensure that the partitions are done with their bootstrapping (not just the placement via controller, but actually started the partitions and Raft groups). I'm looking at `controller_api::wait_for_topic` -- does that do the same thing?

mmaslankaprv replied:
yes, exactly i think `wait_for_topic` is what you actually need

andrwng replied:
Thanks for the pointer, done

andrwng commented on src/v/cluster/cloud_metadata/cluster_recovery_backend.cc:
```diff
@@ -248,7 +252,35 @@ ss::future<cluster::errc> cluster_recovery_backend::do_action(
+        auto synced_term = _raft0->term();
+        while (true) {
+            if (_raft0->term() != synced_term) {
+                co_return cluster::errc::not_leader;
```
That's right, or another node will become leader and try waiting again

andrwng commented on src/v/cluster/cloud_metadata/cluster_recovery_backend.cc:
```diff
@@ -400,12 +407,8 @@ ss::future<> cluster_recovery_backend::recover_until_term_change() {
     ss::abort_source term_as;
     _term_as = term_as;
     auto reset_term_as = ss::defer([this] { _term_as.reset(); });
-    auto synced_term_opt = co_await _recovery_manager.sync_leader(term_as);
-    if (!synced_term_opt.has_value()) {
-        co_return;
-    }
-    auto synced_term = synced_term_opt.value();
-    if (!_recovery_table.local().is_recovery_active()) {
```
Yeah `sync_leader` under the hood checks that the term before is the same as the term after, so the equality check was previously implicit

andrwng commented on src/v/cluster/types.h:
```diff
@@ -113,6 +113,63 @@ struct allocate_id_reply
 
+struct reset_id_allocator_request
+  : serde::envelope<
+      allocate_id_request,
```
Huh, interesting that this even works

andrwng replied:
Done

andrwng commented on src/v/cluster/service.cc:
```diff
@@ -679,6 +679,27 @@ ss::future<transfer_leadership_reply> service::transfer_leadership(
+    auto highest_pid = co_await _partition_manager.map_reduce0(
+      [](const partition_manager& pm) {
+          model::producer_id pid{};
+          for (const auto& [_, p] : pm.partitions()) {
+              pid = std::max(pid, p->highest_producer_id());
+          }
```
It's a good question, and the answer is pretty much the same as the question I posed in the cover letter:

>Why can't we just upload a single producer ID from the id_allocator when uploading metadata? Segment uploads are not coordinated by any central entity. This means that if there were a central uploader that queried the highest producer ID from the id_allocator, there would be a strong chance that new transactional segments get initialized and uploaded before the next ID is included in uploaded metadata.

mmaslankaprv replied:
if i understand that correctly nothing will prevent a new request  allocating new producer id to be processed by any of the partition after this full scan. Would that be the same as in the case we check only the `id_allocator` for the highest producer_id ?

andrwng replied:
>Would that be the same as in the case we check only the id_allocator for the highest producer_id ?

Right, exactly. Because all uploads are disjoint with no stop-the-world kind of sync point, there's always the chance for a window where more producer_ids may be used before the next id_allocator upload (whether that's uploading/scanning the entire partition, or just snapshotting the lastest id state of `id_allocator_stm`).

dotnwat replied:
>pretty much the same as the question I posed in the cover letter:

ahh missed th

andrwng commented:
Latest changes:
- Rebased (archival changes pulled out into https://github.com/redpanda-data/redpanda/pull/15056)
- Removed the "wait for leaders" method and used the existing `wait_for_topics()`
- Fixed bad envelope definition

bharathv commented:
@andrwng What is the conclusion on local segments getting populated during recovery?

andrwng commented:
> @andrwng What is the conclusion on local segments getting populated during recovery?

Sorry for not circling back here: local segments will still be downloaded specifically when remote reads are disabled on the restored topic. There'd been discussion about disabling that behavior because it has shown to be error prone, but it hasn't materialized (generally folks seem on board though, but disabling would be just a stopgap fix, vs fixing the edge cases, see https://github.com/redpanda-data/core-internal/issues/751).
